### DISCUSSION TITLE
Automatically create a new bunch with all open apps and their application state?

### DISCUSSION THREAD

**Discussion by jayhorowitz** [ttscoff/bunch-docs#164] [Q&A] [answered]
I found Bunch in my search for an app that would allow me to save and restore macOS 'sessions' or workspaces — down to the application state. For example,  I currently have ~20 documents open in Preview, but I am almost done for the day with the associated project. I want to one-click create a 'Preview Bunch' that saves all ~20 Preview tabs and windows to a Bunch that I could restore at any time in the future. (Or a Preview and Finder Bunch, or all open apps, etc.)

macOS saved application states are not a robust solution for this, as I want to be able to use the same applications with different files open (i.e. a different session / workspace / bunch) before returning to the most recently saved set.

I see that there is a "New Bunch With Open Apps" menu item, but that only saves the apps that are open, not their application state. Am I in the right place for a solution?

🙏

**Comment by ttscoff** [comment, marked as answer]
There's no universal way for Bunch to know what windows are open in any 
given app. You could probably use AppleScript to cycle through and 
return a list of windows/tabs and their filenames in Preview, but the 
same script wouldn't necessarily work in another app. And knowing window 
state, e.g. which windows in which position and at what size is also not 
feasible.

In short, knowing what apps are open is easy, and using scripting to 
determine what files are open is possible for some apps, but there's not 
a universal way for Bunch to do so. It's up to the user to list out the 
files they want opened in any given application. (A task which can be 
made simpler with a little AppleScripting when there are a significant 
number of files to open).

-Brett

**Reply by jayhorowitz** [reply]
Bummer. I think you'd have a lot of interest if you offer that functionality. On the bright side — and I say this as someone who has no idea what they are talking about — I think it may be possible. See [here](https://forum.keyboardmaestro.com/t/finding-the-file-path-for-the-front-document-in-the-front-application/7095).

I am projecting here, but I believe that much of your potential user base is looking for a GUI-based solution precisely because they want to avoid 'a little AppleScripting.' The moment programming becomes necessary, an overwhelming majority of the computer-using world are no longer potential customers. An unfortunate reality, but reality.

It's easy to say that "it's up to the user to list out the files they want opened in any given application," but the people who are looking for, essentially, a macOS session manager, are precisely the people for whom that is not a reasonable ask due to the size or complexity of their would-be-Bunch. Just my two cents — I wish you success!

**Comment by ttscoff** [comment]
First, I do not have customers, and I don't currently have any intention 
of making Bunch a commercial app. Bunch is free app and I gain nothing 
by catering to any specific audience. Nobody is paying me to solve a 
problem like this.

I appreciate the feedback, but Bunch is already a scripting language of 
sorts, (and not a GUI-based solution as you imply), and expecting a 
certain level of technical skill (such as the occasional fairly basic 
AppleScript) from users doesn't seem beyond reason to me.

I'm not trying to be mean, truly, but you're working off of a couple 
assumptions that I consider to be faulty.

This is not to say I would never consider this functionality, I just 
want to be clear that it's an extensive undertaking that I would be 
doing, essentially, for free.

**Reply by jayhorowitz** [reply]
Your reply is not mean at all. It's honest, and your expectations are hardly unreasonable. Thank you for your setting me straight.

Moving on from my mistaken assumptions, I'll throw out there that I am willing to pay a lot more than 'app prices' for the functionality I described. Particularly if it remains open-source.

If you want to explore the possibility of developing that functionality not-for-free, let me know.
